### PR TITLE
[PR-1251] Article about the state of doctests in Haskell

### PR DESCRIPTION
**Pull request by nalkuatov**
The first draft is ready for review.

### CODE REVIEW

File: the-state-of-doctests-in-haskell/article.md [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+also helps to get the most out of your testing efforts.
+
+For the sake of simplicity and convenience, we won't cover the markup used to define a doctest.
+Instead, we'll try to go over various libraries the community uses to run the doctest examples.
+Particularly, we'll describe a brand new [`cabal-docspec`](https://github.com/phadej/cabal-extras/blob/master/cabal-docspec/MANUAL.md)
+library in a little bit more details.
+
+### Creating a demo project containing doctests
+
+Let's create a project with short and simple haddock documentation containing doctests:
```
Comment (NaeosPsy) [inline the-state-of-doctests-in-haskell/article.md]: Reasoning for creating this file should be added here.

File: the-state-of-doctests-in-haskell/article.md [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+        -- |
+        -- >>> bar
+        -- "bar"
+        bar :: String
+        bar = "bar"
+
+      ```
+
+4. Next, we'll be launching those doctest examples using various libraries in the following sections.
```
Comment (NaeosPsy) [inline the-state-of-doctests-in-haskell/article.md]: If you add reasoning on top, it means this can be deleted.

File: the-state-of-doctests-in-haskell/article.md [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+### Libraries walkthrough
```
Comment (NaeosPsy) [inline the-state-of-doctests-in-haskell/article.md]:
```suggestion
### Library walkthrough
```

File: the-state-of-doctests-in-haskell/article.md [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    `cd doctests-demo`
+
+    `doctest src`
+
+3. The command above should output something like the following:
+
+    `Examples: 2  Tried: 2  Errors: 0  Failures: 0`
+
+While being an awesome software, the library has some limitations and drawbacks.
+Firstly, it has a dependency on [ghc as a library](https://wiki.haskell.org/GHC/As_a_library) - it's very likely that something eventually breaks down when
```
Comment (NaeosPsy) [inline the-state-of-doctests-in-haskell/article.md]: GHC

File: the-state-of-doctests-in-haskell/article.md [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+you try to switch the project to a new compiler version.
+Secondly, when used via [`stack`](https://docs.haskellstack.org/en/stable/README/), one might need to create an additional test suite to bring the project
+dependencies into the scope:
+
+* Let's edit our doctest examples a little (don't forget to add `aeson` and `text` packages to the list of project dependencies):
```
Comment (NaeosPsy) [inline the-state-of-doctests-in-haskell/article.md]: I would use an HTML spoiler tag with the name "example"

File: the-state-of-doctests-in-haskell/article.md (hunk 4 above)
Comment (NaeosPsy) [inline the-state-of-doctests-in-haskell/article.md]: If something is firstly, secondly, thirdly, it can probably be a numbered list.

Comment (NaeosPsy) [reply the-state-of-doctests-in-haskell/article.md]: Here and elsewhere

File: the-state-of-doctests-in-haskell/article.md [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+
+To put it simply, doctests are pieces of text within [haddock comments](https://www.haskell.org/haddock/) that look like interactive Haskell sessions.
+The idea to search for and execute those sessions all over Haskell modules actually comes from the awesome
+[`doctest`](https://en.wikipedia.org/wiki/Doctest) module, which is included in the [`Python programming language`](https://en.wikipedia.org/wiki/Python_(programming_language)).
+
+Documenting software is extremely challenging.
+The doctests concept makes the documentation process not only pleasant and effective - it actually allows Haskell functions to speak for themselves - but
+also helps to get the most out of your testing efforts.
+
+For the sake of simplicity and convenience, we won't cover the markup used to define a doctest.
```
Comment (NaeosPsy) [inline the-state-of-doctests-in-haskell/article.md]: I think it would be more convenient if that would be covered. 

After I read this post, I want to know exactly how to define doctests and which library I want to use for this purpose.

File: the-state-of-doctests-in-haskell/article.md [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+Instead, we'll try to go over various libraries the community uses to run the doctest examples.
```
Comment (NaeosPsy) [inline the-state-of-doctests-in-haskell/article.md]: two libraries, according to the article

File: the-state-of-doctests-in-haskell/article.md [hunk 8]
```diff
@@ -0,0 +1,253 @@
+# The state of doctests in Haskell
```
Comment (NaeosPsy) [inline the-state-of-doctests-in-haskell/article.md]: State of Doctests in Haskell

File: the-state-of-doctests-in-haskell/article.md [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+
```
Comment (NaeosPsy) [inline the-state-of-doctests-in-haskell/article.md]: You need an intro before this heading: it is not good to start an article with a second heading.

File: the-state-of-doctests-in-haskell/article.md [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+    `cd doctests-demo`
+
+    `doctest src`
+
+3. The command above should output something like the following:
+
+    `Examples: 2  Tried: 2  Errors: 0  Failures: 0`
+
+While being an awesome software, the library has some limitations and drawbacks.
```
Comment (NaeosPsy) [inline the-state-of-doctests-in-haskell/article.md]: Some subsections like 

Cons of using `doctest` 
Pros of using `cabal-docspec` 

would add additional strucutre.

File: the-state-of-doctests-in-haskell/article.md (hunk 5 above)
Comment (NaeosPsy) [reply the-state-of-doctests-in-haskell/article.md:84]: https://www.w3schools.com/tags/tag_details.asp

File: the-state-of-doctests-in-haskell/article.md [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+
+7. Now, the `cabal-docspec` command should succeed:
+
+      ```
+      Total:         2; Tried:    2; Skipped:    0; Success:    2; Errors:    0; Failures    0
+      Examples:      2; Tried:    2; Skipped:    0; Success:    2; Errors:    0; Failures    0
+      ```
+
+Thanks for reading!
+Subscribe to our newsletter to stay tuned - `cabal-docspec` definitely needs a dedicated post covering a little more about its configuration and internals!
```
Comment (NaeosPsy) [inline the-state-of-doctests-in-haskell/article.md]: Wouldn't mind if the info was here.

File: the-state-of-doctests-in-haskell/article.md (hunk 2 above)
Comment (heitor-lassarote) [reply the-state-of-doctests-in-haskell/article.md]: Delete the `4.` here maybe? It looks like a new step is going to be here, but there's a new section instead.

File: the-state-of-doctests-in-haskell/article.md [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+
+1. Install the library via `stack`:
+
+    `stack install doctest`
+
+2. Go to the project's root directory and use the library's executable:
+
+    `cd doctests-demo`
+
+    `doctest src`
```
Comment (heitor-lassarote) [inline the-state-of-doctests-in-haskell/article.md]: I'm not sure why, but I get an error here:

```
❯ doctest src
doctest: doctest: can't find a package database at /home/heitor/.cabal-sandbox/x86_64-linux-ghc-8.6.5-packages.conf.d
```

I got around it by adding `doctest` as a project dependency and running `cabal run doctest src`. Simply using `stack exec -- doctest src` also works and avoids the dependency.

File: the-state-of-doctests-in-haskell/article.md (hunk 3 above)
Comment (heitor-lassarote) [reply the-state-of-doctests-in-haskell/article.md]: @NaeosPsy, shouldn't it be `### Library walkthrough`?

File: the-state-of-doctests-in-haskell/article.md [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+    ```haskell
+        -- src/Sample.hs
+
+        {-# LANGUAGE OverloadedStrings #-}
+        {-# LANGUAGE TemplateHaskell   #-}
+
+        module Sample where
+
+        import Data.Text
+        import Data.Aeson
```
Comment (heitor-lassarote) [inline the-state-of-doctests-in-haskell/article.md]: Nitpick: could you sort it alphabetically, by putting `Data.Text` after `Data.Aeson.TH`?

File: the-state-of-doctests-in-haskell/article.md [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+
+        -- |
+        -- >>> encode favourite
+        -- "{\"title\":\"One-Punch Man\",\"rating\":8.9}"
+        --
+        favourite :: Anime
+        favourite = Anime "One-Punch Man" 8.9
+    ```
+
+* Now, if we try to run the doctests using `doctest src` command, we'll fail with the following message:
```
Comment (heitor-lassarote) [inline the-state-of-doctests-in-haskell/article.md]: It works fine with `stack exec -- doctests src`, but I assume this is a problem on my machine. The tests below run without problems, though.

File: the-state-of-doctests-in-haskell/article.md [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+        foo :: Integer
+        foo = 42
+
+        -- |
+        -- >>> bar
+        -- "bar"
+        bar :: String
+        bar = "bar"
+
+      ```
```
Comment (heitor-lassarote) [inline the-state-of-doctests-in-haskell/article.md]: Could you here and in other parts of this article remove the leading indentation? For users that want to copy and paste this code, they will also need to delete the extra spaces.

File: the-state-of-doctests-in-haskell/article.md (hunk 4 above)
Comment (heitor-lassarote) [reply the-state-of-doctests-in-haskell/article.md]: +1 on this. When I got to the "thirdly" I was confused trying to find where were the "firstly" and "secondly" or what it was referring to.

File: the-state-of-doctests-in-haskell/article.md [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+
+    data Anime =
+      Anime { title  :: Text
+            , rating :: Double
+            }
+
+    $(deriveJSON defaultOptions ''Anime)
+
+    -- |
+    -- >>> import Data.Aeson
```
Comment (heitor-lassarote) [inline the-state-of-doctests-in-haskell/article.md]: It took me a minute eyeballing this code to check what is different from the previous one. I thought some top-level import (at the top of the file) changed, but it was this. I think you can do something like this instead:

```diff
    -- |
+   -- >>> import Data.Aeson
    -- >>> encode favourite
    -- "{\"title\":\"One-Punch Man\",\"rating\":8.9}"
    --
    favourite :: Anime
    favourite = Anime "One-Punch Man" 8.9
```

Just providing this definition of `favourite` should be enough rather than the entire module, but if you prefer to keep everything, then it's also fine, just make sure to make it evident where it's different.

File: the-state-of-doctests-in-haskell/article.md (hunk 11 above)
Comment (heitor-lassarote) [reply the-state-of-doctests-in-haskell/article.md]: I actually got to this part and read the "definitely needs a dedicated post" and thought, will we do it or not? I think you should reword it to something like "Subscribe to our newsletter to stay tuned - on the next part of this series, we will post about the configuration and internals of `cabal-docspec`!"

File: the-state-of-doctests-in-haskell/article.md [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+
+Thirdly, the library seems to be slow enough for large-scale projects.
+One of the reasons causing the performance issue is that the library reloads the sources between each doctests' example group.
+This is done to avoid those example groups depending on each other.
+Using the `--fast` flag helps the situation a little, but this is not an option for the recent GHC versions.
+You can learn more about this in the [library's readme](https://github.com/sol/doctest/blob/main/README.md).
+
+### `cabal-docspec`
+
+[`cabal-docspec`](https://github.com/phadej/cabal-extras/blob/master/cabal-docspec/MANUAL.md) library might not seem to be ready to be used at first glance, but
```
Comment (heitor-lassarote) [inline the-state-of-doctests-in-haskell/article.md]: Why "might not seem to be ready to be used at first glance"? It looks fine to me.

File: the-state-of-doctests-in-haskell/article.md [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+        src/Sample.hs:9:1: error:
+            Could not find module ‘Data.Aeson’
+            Perhaps you meant Data.Version (from base-4.14.3.0)
+            Use -v (or `:set -v` in ghci) to see a list of the files searched for.
+          |
+        9 | import Data.Aeson
+          | ^^^^^^^^^^^^^^^^^
+    ```
+
+* If you're using `stack`, the workaround is to create an additional test suite:
```
Comment (heitor-lassarote) [inline the-state-of-doctests-in-haskell/article.md]: It's likely the user is using Stack at this point, you already instructed to use some Stack commands earlier in the post.

File: the-state-of-doctests-in-haskell/article.md [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+The library uses the compiled code - it makes it a lot faster than the alternatives.
+Also, the library doesn't depend on ghc as a library, so it's much more resilient to ghc version changes.
+All of the above makes it more than a viable alternative.
+
+However, the library seems to be closely bound to [`cabal-install`](https://wiki.haskell.org/Cabal): it uses `cabal-install` generated metadata - the `plan.json` file.
+
+Let's consider using `cabal-docspec` to run the doctests in our `doctests-demo` project:
+
+1. First, setup `cabal-install` and the global compiler alongside with all of the auxiliary tooling.
+We suggest using [ghcup](https://www.haskell.org/ghcup/) to achieve this.
```
Comment (heitor-lassarote) [inline the-state-of-doctests-in-haskell/article.md]: Running `ghcup tui` has no `cabal-install`, just `cabal`. I was a bit confused here and just decided to skip to the next step. Also, why "all of the auxiliary tooling"? The user definitely won't need, for example, HLS for this. Also, remember you instructed to use Stack earlier, so if the user didn't install Stack using `ghcup` (like my case), installing it again might not be a good idea.

**Review by heitor-lassarote — CHANGES_REQUESTED** [review]
Nice post. It felt more like a "doctests tutorial" rather than "state of doctests" to me, which sounds like you'd be comparing libraries, their popularity, pros, and cons, etc, but mostly it was guiding the user on how to configure such tests, so I'd rethink the title to better match the contents.
I left some comments as someone new to doctests reading this, I hope they make sense and will add more value to this post.

File: the-state-of-doctests-in-haskell/article.md (hunk 3 above)
Comment (nalkuatov) [reply the-state-of-doctests-in-haskell/article.md]: Shouldn't it be in plural form since two libraries are covered?

File: the-state-of-doctests-in-haskell/article.md (hunk 17 above)
Comment (nalkuatov) [reply the-state-of-doctests-in-haskell/article.md]: I've rephrased the sentence a little, hope it makes more sense now

**Comment by nalkuatov** [comment]
I've applied your suggestions except a proper intro, the markup and technical details (regarding running the examples) - I'll add these very soon

File: the-state-of-doctests-in-haskell/article.md (hunk 3 above)
Comment (NaeosPsy) [reply the-state-of-doctests-in-haskell/article.md]: Yes, Heitor confused me. 

I would actually suggest just to write "Doctest libraries in Haskell"

File: the-state-of-doctests-in-haskell/article.md [hunk 20]
```diff
@@ -63,191 +63,199 @@ Follow the steps below in order to launch the doctest examples in our `doctests-
...
-                }
+import Data.Aeson
+import Data.Aeson.TH
+import Data.Text
 
-        $(deriveJSON defaultOptions ''Anime)
+data Anime =
+    Anime { title  :: Text
+        , rating :: Double
+        }
```
Comment (heitor-lassarote) [inline the-state-of-doctests-in-haskell/article.md]: Nitpick: properly align the indentation here.

File: the-state-of-doctests-in-haskell/article.md [hunk 21]
```diff
@@ -63,191 +63,199 @@ Follow the steps below in order to launch the doctest examples in our `doctests-
...
-    import Data.Aeson
-    import Data.Aeson.TH
+```haskell
 
-    data Anime =
-      Anime { title  :: Text
-            , rating :: Double
-            }
+-- |
+-- >>> import Data.Aeson
```
Comment (heitor-lassarote) [inline the-state-of-doctests-in-haskell/article.md]: I still think this could be made more evident. Two solutions that would work nicely:
1. Right after the "Workaround", mention that you just added this import, or
2. Change syntax highlighting from "haskell" to "diff" and add a + on the first column of this line.

File: the-state-of-doctests-in-haskell/article.md (hunk 11 above)
Comment (heitor-lassarote) [reply the-state-of-doctests-in-haskell/article.md]: Cool. Now we just can't forget to actually create the new post in the future. ;)

Just remember to put the actual info on how to subscribe as @NaeosPsy mentioned.

File: the-state-of-doctests-in-haskell/article.md (hunk 5 above)
Comment (heitor-lassarote) [reply the-state-of-doctests-in-haskell/article.md]: I'm not sure I agree with these changes, the examples are short and I think they're best left auto-expanded.

Comment (NaeosPsy) [reply the-state-of-doctests-in-haskell/article.md]: Yeah, facts. No idea why I commented that.

File: the-state-of-doctests-in-haskell/article.md (hunk 19 above)
Comment (nalkuatov) [reply the-state-of-doctests-in-haskell/article.md]: I suppose running `ghcup install cabal` installs `cabal-install`, as the `cabal-install` package contains the `cabal` executable. The user guide [says](https://www.haskell.org/ghcup/guide/) that they're the same thing :) 
> \# install cabal-install
> ghcup install cabal

File: the-state-of-doctests-in-haskell/article.md (hunk 14 above)
Comment (nalkuatov) [reply the-state-of-doctests-in-haskell/article.md]: I suspect this is because `stack exec` [provides the same environment](https://docs.haskellstack.org/en/stable/GUIDE/#stack-exec) that was used to build the project. 
The idea here was to show the `doctest`'s limitation as of a standalone tool

File: the-state-of-doctests-in-haskell/article.md (hunk 12 above)
Comment (nalkuatov) [reply the-state-of-doctests-in-haskell/article.md]: Looks like this is a problem on your machine ^^

I suppose this can be resolved by slightly changing the ghc options, the `-package-db`

File: introduction-to-doctests-in-haskell/article.md [hunk 22]
```diff
@@ -0,0 +1,289 @@
...
+[`doctest`](https://en.wikipedia.org/wiki/Doctest) module.
+
+Documenting software is extremely challenging.
+Doctests make the documentation process not only pleasant and effective – it actually allows Haskell functions to speak for themselves – but
+also helps to get the most out of your testing efforts.
+
+## How to define a doctest?
+
+
+Let's look at an example of a doctest.
```
Comment (NaeosPsy) [inline introduction-to-doctests-in-haskell/article.md]: Are there any other doctest libraries? Do they all follow the same syntax?

File: introduction-to-doctests-in-haskell/article.md [hunk 23]
```diff
@@ -0,0 +1,289 @@
...
+
+```haskell
+-- |
+-- >>> putStrLn ""
+-- <BLANKLINE>
+```
+
+These are some basic markup requirements:
+
+* Every doctest example should be placed within a valid piece of Haddock documentation, that is followed by either special `-- |` or `{- |` syntaxes.
```
Comment (NaeosPsy) [inline introduction-to-doctests-in-haskell/article.md]: Is the `-- |` or `{-|` just for signifying Haddock documentation?

File: introduction-to-doctests-in-haskell/article.md [hunk 24]
```diff
@@ -0,0 +1,289 @@
...
+Documenting software is extremely challenging.
+Doctests make the documentation process not only pleasant and effective – it actually allows Haskell functions to speak for themselves – but
+also helps to get the most out of your testing efforts.
+
+## How to define a doctest?
+
+
+Let's look at an example of a doctest. 
+
+```haskell
```
Comment (NaeosPsy) [inline introduction-to-doctests-in-haskell/article.md]: Want to perhaps take something from Haskell source? 

Like:
```haskell
-- | @const x@ is a unary function which evaluates to @x@ for all inputs.
--
-- >>> const 42 "hello"
-- 42
--
-- >>> map (const 42) [0..3]
-- [42,42,42,42]
```
Here, the lack of comment and the <BLANKLINE> seems a bit confusing to me, personally.

File: introduction-to-doctests-in-haskell/article.md (hunk 23 above)
Comment (NaeosPsy) [reply introduction-to-doctests-in-haskell/article.md]: Checked Haddock docs myself. ;D

File: introduction-to-doctests-in-haskell/article.md (hunk 24 above)
Comment (NaeosPsy) [reply introduction-to-doctests-in-haskell/article.md]: I've added this particular piece a bit higher, let me know if you don't like that.

File: introduction-to-doctests-in-haskell/article.md [hunk 25]
```diff
@@ -0,0 +1,336 @@
...
+
+#### Why you should use `cabal-docspec`
+
+Here are the reasons why you might want to use `cabal-docspec` for your project: 
+
+* The library is a lot faster than the alternatives because it uses compiled code.
+
+* The library doesn't depend on GHC as a library, so it's much more resilient to GHC version changes.
+
+* The library doesn't require recompiling the source code if it's only about writing doctests – this saves additional time for developers and
```
Comment (NaeosPsy) [inline introduction-to-doctests-in-haskell/article.md]:
```suggestion
* The library doesn't require recompiling the source code if you only change the doctests – this saves additional time for developers and
```

Is this what you meant?

Comment (nalkuatov) [reply introduction-to-doctests-in-haskell/article.md]: Yup, exactly

File: introduction-to-doctests-in-haskell/article.md (hunk 24 above)
Comment (nalkuatov) [reply introduction-to-doctests-in-haskell/article.md]: I do agree with the changes :)

File: introduction-to-doctests-in-haskell/article.md (hunk 22 above)
Comment (nalkuatov) [reply introduction-to-doctests-in-haskell/article.md]: There're a few more libraries. But, AFAIK, they do follow the same syntax

File: introduction-to-doctests-in-haskell/article.md [hunk 26]
```diff
@@ -0,0 +1,336 @@
+# Introduction to Doctests in Haskell
+
+Documenting software can be challenging, but it doesn't always need to be so.
+
+In this article, we'll introduce a concept – doctests – that makes the documentation process pleasant and effective.
```
Comment (heitor-lassarote) [inline introduction-to-doctests-in-haskell/article.md]: Nitpick: I think this text flows a bit better in my mind:

In this article, we'll introduce doctests: a concept that makes the documentation process pleasant and effective.

**Review by heitor-lassarote — APPROVED** [review]
Only one nitpick, but overall it looks pretty good. Nice job :+1:

File: introduction-to-doctests-in-haskell/article.md (hunk 26 above)
Comment (NaeosPsy) [reply introduction-to-doctests-in-haskell/article.md]: Agree.
